### ISSUE TITLE
[inputs.ping] Error when trying to lookup hosts

### ISSUE THREAD

**Issue by oijkn** [influxdata/telegraf#7386]
### Relevant telegraf.conf:

```toml
[global_tags]


# Configuration for telegraf agent
[agent]
    interval = "60s"
    debug = false
    hostname = "CentOS8"
    round_interval = true
    flush_interval = "10s"
    flush_jitter = "0s"
    collection_jitter = "0s"
    metric_batch_size = 1000
    metric_buffer_limit = 10000
    quiet = false
    logfile = "/var/log/telegraf/telegraf.log"
    omit_hostname = false

###############################################################################
#                                  OUTPUTS                                    #
###############################################################################

[[outputs.influxdb]]
    urls = ["http://192.168.1.19:8086"]
    skip_database_creation = true
    database = "Ping"
    timeout = "10s"
    retention_policy = ""

###############################################################################
##                                  INPUTS                                     #
################################################################################

# Ping given url(s) and return statistics
[[inputs.ping]]
  interval = '300s'
  urls = ['host01']
  count = 5
  ping_interval = 1.0
  timeout = 3.0
  deadline = 3
  method = 'native'
  interface = 'eth0'
  [inputs.ping.tags]
     plaque = '01'
     equipement = 'bng'

[[inputs.ping]]
  interval = '300s'
  urls = ['host02']
  count = 5
  ping_interval = 1.0
  timeout = 3.0
  deadline = 3
  method = 'native'
  interface = 'eth0'
  [inputs.ping.tags]
     plaque = '02'
     equipement = 'bng'

[[inputs.ping]]
  interval = '300s'
  urls = ['host03']
  count = 5
  ping_interval = 1.0
  timeout = 3.0
  deadline = 3
  method = 'native'
  interface = 'eth0'
  [inputs.ping.tags]
     plaque = '03'
     equipement = 'bng'

[[inputs.ping]]
  interval = '300s'
  urls = ['host04']
  count = 5
  ping_interval = 1.0
  timeout = 3.0
  deadline = 3
  method = 'native'
  interface = 'eth0'
  [inputs.ping.tags]
     plaque = '04'
     equipement = 'bng'

[[inputs.ping]]
  interval = '300s'
  urls = ['host05']
  count = 5
  ping_interval = 1.0
  timeout = 3.0
  deadline = 3
  method = 'native'
  interface = 'eth0'
  [inputs.ping.tags]
     plaque = '05'
     equipement = 'bng'
```

### System info:

```
# uname -a
Linux TIG 4.18.0-147.8.1.el8_1.x86_64 #1 SMP Thu Apr 9 13:49:54 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
# telegraf --version
Telegraf 1.14.1 (git: HEAD 33dd867d)
# influx --version
InfluxDB shell version: 1.8.0
```

### Steps to reproduce:

1. Setup ping telegraf and ping plugin with single host per plugin -> 5 hosts for testing
2. Ping a host that is or not resolvable using DNS server

### Expected behavior:

No such host if it's not resolvable or if it can be reached wait for the response from the DNS server.

### Actual behavior:

```
2020-04-22T15:35:40Z E! [inputs.ping] Error in plugin: lookup host1 on dns02.server:53: no such host
2020-04-22T15:37:14Z E! [inputs.ping] Error in plugin: lookup host2 on dns01.server:53: read udp 192.168.6.128:36283->dns01.server:53: i/o timeout
2020-04-22T15:37:14Z E! [inputs.ping] Error in plugin: lookup host1 on dns02.server:53: read udp 192.168.6.128:53474->dns02.server:53: i/o timeout
2020-04-22T15:37:14Z E! [inputs.ping] Error in plugin: lookup host3 on dns01.server:53: read udp 192.168.6.128:39761->dns01.server:53: i/o timeout
2020-04-22T15:37:38Z E! [inputs.ping] Error in plugin: lookup host2 on dns02.server:53: no such host
2020-04-22T15:37:38Z E! [inputs.ping] Error in plugin: lookup host3 on dns02.server:53: no such host
```

### Additional info:

Well I don't really know where to start because I did a lot of tests before creating this issue.
First of all I've already discussed a little bit about it  about it on https://github.com/influxdata/telegraf/pull/6743.

At first I thought that the error could be due to the fact that we are trying to search for so many hosts at the same time or that one or more hosts are not reachable. Later I realized that even if I remove the unreachable host from the list, I still get errors in the log file.

At first I was on the following configuration :

- VirtualBox with CentOS 7 (2vcpu / 4Go RAM)
- Network :
    - eth0 (NAT -> connexion LAN to DNS + Hosts)
    - eth1 (Bridge -> connexion with my Laptop)

With this configuration I had buffer problems with the ARP table, certainly caused to the large number of requests because initially I had +500 hosts in my list. So I decided to check with a different environment.

Second configuration :

- VMware Workstation with CentOS 8 (2vcpu / 4Go RAM)
- Network :
    - eth0 (NAT -> connexion LAN to DNS + Hosts) -> burst mode + DNS forced in settings
    - eth1 (Bridge -> connexion with my Laptop)

I guess VMware is more robust because I didn't have any buffer problem or anything else but the issue with the plugin remains the same :(

When I do a `dig` command of the host the answer is : 37 msec. But if I do a `ping`, it takes at least 3 or 4 seconds before the ping responds.

By modifying the settings of `/etc/resolv.conf` I get better results but that's not it yet...
```
search my.suffix
nameserver dns01.server
nameserver dns02.server

options timeout:1
options attempts:3
options rotate
```

I tried it anyway with the `method = 'exec'` and it seems that the problem does not appear (just once when launching the telegraf service).
On the other hand I see the 's' option for the packet size which is 16. The default packet size is 56, is this a bug ? `/usr/sbin/ping -c 5 -n -s 16 -i 1 -W 3 -w 3 -I eth0 host1`

Thank you for your support and help.

**Comment by oijkn** [comment]
@danielnelson I'm still investigating because I suspect a DNS related problem, either in my CentOS config or server conf (on this side I don't think so).

**Comment by danielnelson** [comment]
It may also be worth trying out the [nightly builds](https://github.com/influxdata/telegraf#nightly-builds), they include #7140 which cut the number of DNS queries in half.  My leading theory is still that we are just making too many DNS queries at once.

**Comment by oijkn** [comment]
In order to avoid problems with DNS queries I only let the NAT interface and commands ping, dig respond correctly.

I installed the latest nightly builds from your link and I still have errors that appear in log file : 
```
2020-04-23T12:35:03Z E! [inputs.ping] Error in plugin: i/o timeout
2020-04-23T12:35:03Z E! [inputs.ping] Error in plugin: i/o timeout
2020-04-23T12:35:03Z E! [inputs.ping] Error in plugin: i/o timeout
2020-04-23T12:35:03Z E! [inputs.ping] Error in plugin: i/o timeout
2020-04-23T12:35:03Z E! [inputs.ping] Error in plugin: i/o timeout
2020-04-23T12:35:03Z E! [inputs.ping] Error in plugin: i/o timeout
2020-04-23T12:35:03Z E! [inputs.ping] Error in plugin: i/o timeout
2020-04-23T12:35:03Z E! [inputs.ping] Error in plugin: i/o timeout
```

### Telegraf info:
```
[root@TIG telegraf.d]# rpm -qi telegraf
Name        : telegraf
Version     : 1.15.0~c8b9cb40
Release     : 0
Architecture: x86_64
Install Date: jeu. 23 avril 2020 13:24:14 CEST
Group       : default
Size        : 71083375
License     : MIT
Signature   : (none)
Source RPM  : telegraf-1.15.0~c8b9cb40-0.src.rpm
Build Date  : jeu. 23 avril 2020 09:13:57 CEST
Build Host  : a08dee28a147
Relocations : / 
Packager    : support@influxdb.com
Vendor      : InfluxData
URL         : https://github.com/influxdata/telegraf
Summary     : Plugin-driven server agent for reporting metrics into InfluxDB.
Description :
Plugin-driven server agent for reporting metrics into InfluxDB.
```

**Comment by oijkn** [comment]
As a continuity of my tests, I installed dnsmasq on the VM in order to have the information stored in cache, which considerably improves performance and reduces errors. For these tests I have rollback on the stable version.

Apart from the DNS problem, I notice that the `percent_packet_loss` does not reflect the reality. I let a ping run in the background with which I have no packet loss and yet the plugin indicates 33% loss.

```
2020-04-23T16:25:03Z 66.66666666666666
2020-04-23T16:30:03Z 33.33333333333333
2020-04-23T16:35:03Z 33.33333333333333
2020-04-23T16:40:03Z 33.33333333333333
2020-04-23T16:45:03Z 33.33333333333333

--- host1 ping statistics ---
447 packets transmitted, 447 received, 0% packet loss, time 3221ms
rtt min/avg/max/mdev = 52.362/55.457/105.731/3.514 ms
```

**Comment by oijkn** [comment]
I increased from 500 to 1500 hosts and with the `collection_jitter option = "1s"` everything seems to work well. I wanted to go to the next step with 2800 hosts and I have errors in the log file again: **i/o timeout**
Like you said, the error could be because we are trying to lookup so many hosts at the same time.

By passing the `collection_jitter option = "5s"` I have no more error but I guess it shifts the ping interval, right?

Another question, would you please tell me what is the size of the ping sent? Thank you

**Comment by danielnelson** [comment]
> I also tried to put the IP addresses directly instead of the hostname in the configuration file to avoid errors in the log file. And with this method I have no more errors !!

This seems to confirm that we are overwhelming DNS.  One way we could fix this is by using a shared resolver that limits concurrent lookups.

> However I notice an incoherency in the percentage of lost packets...

Does this occur only when pinging many hosts or does it still occur with a single host?

> I think an excellent alternative would be to use fping but I haven't found a plugin complementary to Telegraf

You could probably set this up as an `exec` plugin, you would need to write a wrapper script to parse the output and convert it to line protocol or another `data_format`.

**Comment by oijkn** [comment]
> Does this occur only when pinging many hosts or does it still occur with a single host?

It's occur when I try to ping many hosts

> You could probably set this up as an `exec` plugin, you would need to write a wrapper script to parse the output and convert it to line protocol or another `data_format`.

After testing a lot of github and other projects, I can't find anything that can withstand the load of several thousand targets.

I confess that I had already thought of using the exec plugin to reach my goal and in the end while reading your feedback, I said to myself: why not create my own plugin !

I have good bases in code but not in golang so I'm going to grope around.

**Comment by danielnelson** [comment]
We would love to get some help on the native ping code too, especially if you can figure out why the packet loss isn't as expected.

**Comment by oijkn** [comment]
> We would love to get some help on the native ping code too, especially if you can figure out why the packet loss isn't as expected.

I would like to :+1:

For the moment I'm working in Python script with exec plugin who call fping and result is not bad (I'm trying to optimize speed because output in console take to much time)

**Comment by kw16net** [comment]
Hi, 

sorry for the late reply.
Telegraf runs in a Docker container and DNS resolution happens via the Docker DNS server 127.0.0.11 and are forwarded to the Docker host where I have entries in /etc/hosts.
It works like 99,9% of the time and rarley fails. These i/o timeoutes lead to false-postives entries in the database, which is annoying.

Must not be the same problem as oijkn has, only "i/o timeout"'s happen too.
Maybe it was not a good idea to post my problem in this thread, sorry for that.

**Comment by danielnelson** [comment]
I think it is related, if we are to support the pinging of very many hosts it seems that we need to rethink the DNS lookup in this plugin, possibly caching the lookups across queries.

**Comment by sjwang90** [comment]
Fixed in #8679
